### PR TITLE
feat: implement OAuth logout functionality for keyclock to terminate sso session

### PR DESCRIPTION
**Pull request by ZaibanAli**
# Pull Request Checklist

### Note to first-time contributors: Please open a discussion post in [Discussions](https://github.com/open-webui/open-webui/discussions) and describe your changes before submitting a pull request.

**Before submitting, make sure you've checked the following:**

- [X] **Target branch:** Please verify that the pull request targets the `dev` branch.
- [X] **Description:** Provide a concise description of the changes made in this pull request.
- [X] **Changelog:** Ensure a changelog entry following the format of [Keep a Changelog](https://keepachangelog.com/) is added at the bottom of the PR description.
- [X] **Documentation:** Have you updated relevant documentation [Open WebUI Docs](https://github.com/open-webui/docs), or other documentation sources?
- [X] **Dependencies:** Are there any new dependencies? Have you updated the dependency versions in the documentation?
- [X] **Testing:** Have you written and run sufficient tests for validating the changes?
- [X] **Code review:** Have you performed a self-review of your code, addressing any coding standard issues and ensuring adherence to the project's coding standards?
- [X] **Prefix:** To cleary categorize this pull request, prefix the pull request title, using one of the following:
  - **BREAKING CHANGE**: Significant changes that may affect compatibility
  - **build**: Changes that affect the build system or external dependencies
  - **ci**: Changes to our continuous integration processes or workflows
  - **chore**: Refactor, cleanup, or other non-functional code changes
  - **docs**: Documentation update or addition
  - **feat**: Introduces a new feature or enhancement to the codebase
  - **fix**: Bug fix or error correction
  - **i18n**: Internationalization or localization changes
  - **perf**: Performance improvement
  - **refactor**: Code restructuring for better maintainability, readability, or scalability
  - **style**: Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc.)
  - **test**: Adding missing tests or correcting existing tests
  - **WIP**: Work in progress, a temporary label for incomplete or ongoing work

# Changelog Entry

### Description

- Signout doesn't terminate SSO sessions [ISSUE_LINK](https://github.com/open-webui/open-webui/issues/7583)
- Clicking the logout button only deletes the cookie for Open Webui, but does not end the session in Keycloak.
- As a result, users can still access the webui (and any other SSO-enabled service) without having to re-authenticate by simply clicking the button to sign in with the IDP again.

### Added

- Adding id_token to the cookie when the oidc provider signup is True. This is required to cleanup the session when signing out.

### Changed

- Updated the `/signout` endpoint to include logic for cleaning up the oidc session (keycloak) using the `id_token`. (backend/open_webui/apps/webui/routers/auths.py)
- Enhanced the OAuth callback functionality to set an `id_token` cookie when using the OAUTH oidc provider. (backend/open_webui/utils/oauth.py)

### Deprecated

- [List any deprecated functionality or features that have been removed]

### Removed

- [List any removed features, files, or functionalities]

### Fixed

- Resolved an issue where the `signout` function did not correctly handle Keycloak logouts due to missing `id_token`.

### Security

- [List any new or updated security-related changes, including vulnerability fixes]

### Breaking Changes

- **BREAKING CHANGE**: [List any breaking changes affecting compatibility or functionality]

---

### Additional Information

- [ISSUE_7583](https://github.com/open-webui/open-webui/issues/7583)

### Screenshots or Videos

- [Demonstration Video](https://drive.qaiv.com/s/8tkkEHQXYZ7i6Jm)

### CODE REVIEW

**Comment by sreinwald** [comment]
First of all: thanks for taking on the issue, I was going to tackle it today as well.

Some notes:
I believe adding new ENV variables is not necessary. [OIDC RP-Initiated Standard](https://openid.net/specs/openid-connect-rpinitiated-1_0.html) is to publish the `end_session_endpoint` at the OIDC provider's `/.well-known/openid-configuration` API endpoint, which we get from `OPENID_PROVIDER_URL`. 

It's probably cleaner to check if the OIDC provider supports the OIDC RP-Initiated Standard by supplying the `end_session_endpoint` and then call that, rather than have the user supply two new ENV variables, and do individual checks for each OIDC provider, as your implementation only works for Keycloak right now.

**Comment by ZaibanAli** [comment]
@sreinwald thanks for this information, i didn't knew about this. I agree this makes more sense since this is more elegant and can be used for other providers.

Any idea how to tackle the issue with the specific query parameter i.e. `id_token_hint` required by logout endpoint of keycloak?

**Comment by ZaibanAli** [comment]
@sreinwald I have now removed the additional variable, can you please review again.

File: backend/open_webui/apps/webui/routers/auths.py [hunk 1]
```diff
@@ -3,6 +3,7 @@
 import time
 import datetime
 import logging
+import httpx
```
Comment (tjbck) [inline backend/open_webui/apps/webui/routers/auths.py:6]: should be implemented with `aiohttp`

Comment (ZaibanAli) [reply backend/open_webui/apps/webui/routers/auths.py]: I have updated the code to use aiohttp instead of httpx.

**Comment by tjbck** [comment]
Looks great, Thanks!

**Comment by TheDropZone** [comment]
@ZaibanAli 
So, I've deployed this OAuth functionality alongside a keycloak instance, and am running into some issues with the cookie deletion when open-webui and keycloak live on different domains.
- openwebui01-domain.com
- kc01-domain.com

The cookies (token, oauth_id_token) live and are set on the open-webui domain (openwebui01-domain.com), but the signout url from keycloak (and the openid_config) is on kc01-domain.com. So, it seems that when the signout endpoint requests to delete the cookies (living on openwebui01) but then redirects to kc01-domain for signout, the cookies don't get deleted (due to cookie domain mismatch). Because I'm redirected to the kc01-domain for the signout, the cookies dont exist at that domain to be deleted, and token and oauth_id_token live forever.

Has anyone else seen this issue, or have ideas on the proper way to handle this?

**Comment by mig5** [comment]
Why is this only possible if `ENABLE_OAUTH_SIGNUP` is set to true?

If I've pre-configured some users in Open WebUI with the same attributes that are in my OIDC backend for those users, I should be able to handle SSO logout for those users when logging out of Open WebUI, even if `ENABLE_OAUTH_SIGNUP` is false (I don't want other OIDC users to self-register in Open WebUI).

**Pull request merged**